### PR TITLE
Prevent unnecessary and unexpected calls to capf hooks

### PR DESCRIPTION
**Pull request by joaotavora**
In company--capf-post-completion, we have to access cache directly because,
according to company--capf-data, it is already stale.  Indeed, the freshly
inserted completion has changed point, but the cache is not really stale for
getting the needed :exit-function.  Therefore, we use this cache to prevent a
second, unexpected call to the capf hooks.

* company-capf.el (company--capf-post-completion): Access cache
directly.
(company-capf): Fix closing parens.

### CODE REVIEW

**Comment by joaotavora** [comment]
We could also use a separate cache for the `:exit-function` with different invalidation semantics.  Or transform `company-mode` to use a functional interface and use lexical binding instead of all these scary caches :-)

**Comment by dgutov** [comment]
So... the reason we can do this is because `company--insert-candidate` calls `(company-call-backend 'ignore-case)`? We should probably document this.

If not for this call, `company-sort-by-occurrence` could make the last cached position be somewhere surprising.

**Comment by dgutov** [comment]
> We could also use a separate cache for the :exit-function with different invalidation semantics. Or transform company-mode to use a functional interface and use lexical binding instead of all these scary caches :-)

Could you elaborate on either of that?

**Comment by dgutov** [comment]
To be clear, this is a performance/clarity improvement, not a correctness fix, right?

**Comment by joaotavora** [comment]
> To be clear, this is a performance/clarity improvement, not a correctness fix, right?

No, this is a correctness fix, i.e a bug in company. The only other frontend using capf functions only calls each function once for each full completion attempt, and thus you can write `:exit-function`s that capture the lexical environment. 

With company-mode you can't. This seems. be unintentional because there is for all the operations in the company-capf backend a call to the cached capf data getter. But in the specific case of the post-completion handler this cache will always be invalid, by definition.

> Could you elaborate on either of that?

Sure, once we agree that this is a  bug, or you show me the deterministic cache miss was indeed intentional.

**Comment by joaotavora** [comment]
> So... the reason we can do this is because company--insert-candidate calls (company-call-backend 'ignore-case)? We should probably document this.

Not sure I parse the sentence. If "this that we do" is accessing the cache directly without going through `company--capf-data`, then I think I am just writing code that translates the original intention, which is to get at the cached data that `company--capf-data` presumably calculated just once.  I don't know where `ignore-case` enters into it.

>If not for this call, company-sort-by-occurrence could make the last cached position be somewhere surprising.

I don't understand this. What is the "last cached position" and how is it important to `company--capf-post-completion`?

In this PR, I assumed that _no_ parts of `company--capf-cache` that are useful to `company--capf-post-completion` have been invalidated between the first call to `company--capf-cache` and `company--capf-post-completion`.  Can you show otherwise? What am I missing?

File: company-capf.el [hunk 1]
```diff
@@ -165,11 +165,14 @@
      (plist-get (nthcdr 4 (company--capf-data)) :company-require-match))
     (`init nil)      ;Don't bother: plenty of other ways to initialize the code.
     (`post-completion
-     (company--capf-post-completion arg))
-    ))
+     (company--capf-post-completion arg))))
```
[the remark this answers is not in the document]
Comment (joaotavora) [reply company-capf.el]: This, with all respect, is a bit bizarre since every lisp function is by definition a list.  But I agree it has no place in this PR, so I removed it.

**Comment by joaotavora** [comment]
Earlier, I wrote:

> The only other frontend using capf functions only calls each function once for each full completion attempt, and thus you can write :exit-functions that capture the lexical environment.

Actually, this is slightly misleading.  The only other frontend _i know_ using capf functions, `M-x completion-at-point` in `minibuffer.el` will often call capf hooks more than once, for purposes of serving `completion-in-region-mode-predicate`, which is arguably also a bug (in Emacs proper).

But, nevertheless, it guarantees that the `:exit-function` called is the one returned by the first call to the hook.  This allows for stable captures of the lexical environment.

File: company-capf.el (hunk 1 above)
[the remark this answers is not in the document]
Comment (dgutov) [reply company-capf.el]: Not sure what you mean, I'm just saying that pure-addition or pure-deletions diff hunks are a tiny bit nicer, when one can manage that. So thanks.

[the remark this answers is not in the document]
Comment (joaotavora) [reply company-capf.el]: To continue this bike-shedding, I am saying that you could extend yourreasoning to _any_ lisp function, and make the argument that closing and opening parenthesis should always exist on a line of their own, because everytime you add or remove a form from a lisp function it's a "pure addition" or a "pure deletion". 

This would be bizarre because that's clearly bad lisp style.  _Why_ is that bad lisp style, you ask? I don't know, it's an axiom for me and presumably many others.

[the remark this answers is not in the document]
Comment (dgutov) [reply company-capf.el]: You could, but then it would indeed look very ugly, whereas two closing parens on a separate line look okay to me. So I'd only apply this approach to pieces of code where the shape of the future changes is fairly obvious.

I'm not the original author of that code, though. That was Stefan. I just left that bit as-is.

[the remark this answers is not in the document]
Comment (dgutov) [reply company-capf.el]: BTW, grep the Emacs code base for "^ *)+" sometime.

**Comment by dgutov** [comment]
> which is to get at the cached data that company--capf-data presumably calculated just once

No, the goal is to reduce the number of times it has to be calculated. But to force calculation anyway whenever the buffer, the value of point, or anything else significant changes.

Normally, the effect of this cache on the behavior is minimal because the elements of `completion-at-point-functions` are pretty fast anyway, so you could call them after every user keypress, and they still might not notice the difference. But some of the more carelessly written CAPF functions (e.g. ones that calculate the list of completions right away) get some benefit from this cache. That's why I added it.

(to be continued)

File: company-capf.el (hunk 1 above)
[the remark this answers is not in the document]
Comment (joaotavora) [reply company-capf.el]: > That was Stefan

It could've been McCarthy himself, I still would find it silly :-) But more to our point: doesn't that mean @monnier would be a good reviewer for this change?

> BTW, grep the Emacs code base for "^ *)+" sometime.

Emacs != Lisp style guide ;-)

**Comment by dgutov** [comment]
> What is the "last cached position" and how is it important to company--capf-post-completion?

The value of point for which `company--capf-cache` was calculated.

> The only other frontend i know using capf functions, M-x completion-at-point in minibuffer.el

There are other frontends out there by now (e.g. Counsel and even Auto-Complete), but let's indeed discuss the standard one.

> But, nevertheless, it guarantees that the :exit-function called is the one returned by the first call to the hook.

I don't think it *guarantees* that, it's just how it ends up working for now. But that's the behavior of a frontend that's less feature-rich than Company, or else we wouldn't be using the latter, or discussing this problem here.

So I think it's natural that this package can exercise the CAPF API in ways that you wouldn't be able to anticipate by just looking at what completion-at-point does. So in the end, we can only rely on the CAPF API in abstract.

And as for extra features, take a look at `company--occurrence-predicate`: it calls the `prefix` backend command. Unfortunately, someone might say, the CAPF approach prioritizes flexibility over succinctness, so to calculate the prefix at an arbitrary buffer position we have to call all the elements in `completion-at-point-functions` all over again. Which changes the last cached capf-data in this backend.

Now, this feature is optional and disabled by default, but it's one I really like, and it has its users. So we can't rely on the last cached data staying the same all throughout completion. The reason your patch will probably work fine anyway is that right before the end `company--insert-candidate` calculates it again at the original position, so afterwards `post-completion` can more-or-less safely just access the cached value. This is no strong guarantee, though, and might change in some distant future.

> This allows for stable captures of the lexical environment.

I'm still not sure what this is referring to. Not Emacs Lisp lexical environment, right?

**Comment by dgutov** [comment]
> M-x completion-at-point in minibuffer.el will often call capf hooks more than once, for purposes of serving completion-in-region-mode-predicate, which is arguably also a bug (in Emacs proper).

Last I checked it indeed called hooks several times, and did that after every user command, which might seem extraneous, and never cached the result anywhere. So I think the idea is that the data returned by a CAPF function is only necessarily valid for the given position, right now. Anything else is just a guess. So the caching logic in the `company--capf-data` function is my best guess on the subject.

You are welcome to bring this question up on emacs-devel, though.

**Comment by dgutov** [comment]
Having said all this, I have to ask: is calculating the completion bounds (or the prefix, in Company terms) expensive in Eglot? Do you call the LSP server for it?

**Comment by joaotavora** [comment]
> No, the goal is to reduce the number of times it has to be calculated. But to force calculation anyway whenever the buffer, the value of point, or anything else significant changes.

OK. Is that synonymous with "a new completion attempt happens"?

> Normally, the effect of this cache on the behavior is minimal because the elements of completion-at-point-functions are pretty fast anyway, 

OK, if speed was the only intention then (almost) correctness was a side-effect.  No problem.

> I don't think it guarantees that, it's just how it ends up working for now.

No, I don't agree.  Any frontend that calls a function to receive an object that has functional subobject **and** if that frontend wants to call said functional objects, it should do so for the first instance of the returned object.  You will notice that Emacs's `completion-at-point` adheres, quite sensibly, to this principle, which should be obvious IMO.

**Comment by joaotavora** [comment]
> Having said all this, I have to ask: is calculating the completion bounds (or the prefix, in Company terms) expensive in Eglot? Do you call the LSP server for it?

No, _it's not a question of speed at all_, or at least it's not a question of speed _foremost_. It's a question of correctness.  In Eglot, the function placed in `capf` produces a list with lexical closures. These lexical closures legitimately capture their lexical environment.  But since the function is called more than once **for producing said lexical closures** it's unpredictable exactly what lexical environment they are capturing.

So, in other words: you **can** if you want, call `capf` functions more than once per completion attempt if you really need to. **But** when you are `funcall`ing functions returned by said calls you should call the function returned by the very first call.

**Comment by joaotavora** [comment]
> the function returned by the very first call.

By the very first call in that completion attempt, of course.

File: company-capf.el (hunk 1 above)
[the remark this answers is not in the document]
Comment (monnier) [reply company-capf.el]: Bike shedding, Yay!!  I'll agree with both of you, for a change: closing parens on their own line are ugly, I'm not sure why I left them there.  But "extending the reasoning" doesn't really work: while all "sexps" can be considered as lists, Elisp is mostly made of function calls and macro calls and a large fraction of them have a fixed number of arguments, so adding/removing elements from those lists is not going to happen.

IOW I vote for green with yellow stars.

[the remark this answers is not in the document]
Comment (dgutov) [reply company-capf.el]: Ah hah. All right then. :)

**Comment by monnier** [comment]
I think I see what João is talking about: I don't think describing it as "the very first call" is quite right, and neither is the lexical environment really relevant (the two calls to capf may simply return two different function symbols, with no closures in sight), but indeed, the :exit-function goes together with the completion-table.  If you insert a completion from the completion table, then call capf you might very well receive a different completion table with a different exit-function, so the exit function you should call after inserting a completion is the one you got when you received the original completion table, not the one you get when you re-call capf.

So the problem is that the `company-backend` API doesn't give us the relevant info: when `post-completion` is invoked all we know is what has just been inserted, but the completion data (table and exit-function) which was used to choose this completion is not available to company-capf.  João's patch is a workaround which takes advantage of the existence of the cache to try and recover the relevant completion data, under the assumption that the cache has last been refreshed just before performing the insertion, so it should happen to contain just the data we need.

I don't know whether João's patch is the best solution, but at least I can't think of a better one right now: we could try and de-insert the text, call capf, and then re-insert the text instead, which would avoid the reliance on the cache, but that sounds ugly;  we could also call capf from the position right before the insertion which is a bit less ugly but may sometimes still return incorrect data.

This said, the comment should better explain the problem and make it explicit that this is a workaround.

File: company-capf.el (hunk 1 above)
[the remark this answers is not in the document]
Comment (joaotavora) [reply company-capf.el]: @monnier, well not really, because blah blah blah <dissolves into nothingness> 

Anyway, this wasn't the reason I summoned you, but in the meantime have posted the matter to emacs-devel at Dmitry's suggestion, so let's continue there.

[the remark this answers is not in the document]
Comment (joaotavora) [reply company-capf.el]: Nevermind, I see that you have answered below

**Comment by joaotavora** [comment]
> This said, the comment should better explain the problem and make it explicit that this is a workaround.

I'll take this deal :-) @dgutov ?

**Comment by dgutov** [comment]
All right. I see the issue now, thank you both. But:

> Is that synonymous with "a new completion attempt happens"?

No. See both mentions of `company-sort-by-occurrence` in my comments here as a counter-example.

So relying on the preceding `ignore-case` call, like I suggested before, isn't going to work.

If we want to reliably call the `:exit-function` that was returned by the same completion data that produced the list of completions, we'll need to do something more complex. For instance, to only write to `company--capf-cache` if the backend action is `candidates`, avoid caching the computed data in `prefix`, `require-match` and `ignore-case` actions, and use the saved cache directly in all other actions. Or something along these lines. Which would increase the number of times capf hooks calls are called, but oh well.

I guess a proper fix would be to extend the backend API so that Company keeps a complex "completion data" object returned from the backend instead of just a reference to the said backend during completion. And gives access to the said data to the backends in their action handlers. That would be a bigger change, though.

File: company-capf.el (hunk 1 above)
[the remark this answers is not in the document]
Comment (monnier) [reply company-capf.el]: > @monnier, well not really, because blah blah blah

I think this is rather rude: my dog opined she's offended.  But I guess I could agree to green with blah blah blah, as long as the blah blah blah is using a sufficiently subdued color.  [ Sorry, but I'm stuck on a proof.  ]

**Comment by monnier** [comment]
Maybe the code should be more explicit about it, then:
- Keep the cache as is (i.e. it's only used for performance).
- When `candidates` is called, stash the result in another variable `company-capf--current-completion-data`
- When the exit-function is requested look for it in `company-capf--current-completion-data`.

**Comment by joaotavora** [comment]
>> Is that synonymous with "a new completion attempt happens"?
> Which would increase the number of times capf hooks calls are called, but oh well.

You don't have to sacrifice this if you save the :exit-function and the relevant arguments in another variable that *does* have those semantics. But, in the meantime, you could install my workaround.
That, by the way, would be the first alternative I suggested [here](https://github.com/company-mode/company-mode/pull/845#issuecomment-443896366)

>I guess a proper fix would be to extend the backend API so that Company keeps a complex "completion data" object returned from the backend instead of just a reference to the said backend during completion. And gives access to the said data to the backends in their action handlers. That would be a bigger change, though.

And this would be mostly the second alternative.

**Comment by joaotavora** [comment]
> When candidates is called, stash the result in another variable company-capf--current-completion-data

Yeah, this is also what I think could be done.  I've added this change to the PR.

**Comment by dgutov** [comment]
@monnier

> Keep the cache as is (i.e. it's only used for performance)

Thing is, `(company-capf 'prefix)` is normally very fast (on the order of microseconds), so a few more calls shouldn't be noticeable at all. So we might not even want to use it as a performance optimization, forcing the backend authors to fix their code instead, if necessary.

Two caches also means extra complexity when debugging.

**Comment by dgutov** [comment]
@joaotavora 

> That, by the way, would be the first alternative I suggested here

Cool. But it's not like, given the level of detail, I could just say "yes, let's do that".

More importantly, though, shouldn't we also treat `sorted`, `match`, `meta`, `doc-buffer`, `location` and `annotation` the same way? These are all handled as properties returned together with the completion table, just like `post-completion`.

**Comment by joaotavora** [comment]
> Thing is, (company-capf 'prefix) is normally very fast (on the order of microseconds), so a few more calls shouldn't be noticeable at all. So we might not even want to use it as a performance optimization, forcing the backend authors to fix their code instead, if necessary.

Sorry, but this is just wrong. Why would you deliberately harm users of sub-par code just to prove a point with developers of said code, when you have simpler, 3-loc-long alternatives at your disposal?

> Cool. But it's not like, given the level of detail, I could just say "yes, let's do that".

It was only worth explaining a second solution once you understood the problem, which wasn't immediate ( perhaps due to insufficiently clear explanation from my part)

> shouldn't we also treat sorted, match, meta, doc-buffer, the same way

The current cache strategy handles these cases fine, for intentional or accidental reasons.

 I am not proposing a refactoring/cleanup, rather a bugfix. Please do the latter first using whatever strategy you decide is most promising for the former. I've already proposed two strategies.

**Comment by dgutov** [comment]
> Why would you deliberately harm users of sub-par code just to prove a point with developers of said code, when you have simpler, 3-loc-long alternatives at your disposal?

More like make any performance problems apparent in the default configuration, and not just certain kinds of users. Which could be beneficial to them all.

And I doubt many packages would be affected, considering `completion-at-point` itself makes several calls.

Anyway, there's no urgent need to make this change right now. There's always time for it later.

> It was only worth explaining a second solution once you understood the problem, which wasn't immediate

True. And it seems our concepts of the problem under discussion still don't match exactly.

> The current cache strategy handles these cases fine, for intentional or accidental reasons.

Accidental, if we're speaking strictly. Of course, `annotation-function`, etc, produced in repeated invocations on the same position are likely to work on the list of completions returned first, but it's likely to fail if some of the lexically captured variables are tied to something else than the current buffer contents. Or if there's a hash-table using an `eq` test among them.

So I'd rather not split the fix into several commits. If you don't mind, I'll amend your patch before pushing. Alternatively, I can strip the attribution.

> So does fixing bugs introduced by ignoring it :)

Sorry, I didn't quire catch the meaning here.

**Comment by joaotavora** [comment]
> More like make any performance problems apparent in the default configuration, and not just certain kinds of users. Which could be beneficial to them all.

In the long-run maybe. But in this case is still better in practice to have some kind of optional "strict" mode of operation with which you can pressure developers without punishing users.  

>  And I doubt many packages would be affected, considering completion-at-point itself makes several calls.

I think for good or bad many users assume company is the only frontend, or at least their most important one.

And as far as I could see completion-at-point only does one extra call in the conditions that I described. That one is just for checking the bounds and still a design flaw imo. The many other calls seem to be for updating the completions on partial completion and such, which is of course ok: it counts  as another completion attempt.

> Accidental, if we're speaking strictly.

Anyway, design decisions where you don't repeat actions that you determined that you don't have to repeat are good, even if accidental, and regardless of speed. Of course there are good and bad implementations of such design. A cache is arguably a poor one, an object with closures and a well defined life would be better.


> Of course, annotation-function, etc, produced in repeated invocations on the same position are likely to work on the list of completions returned first, but it's likely to fail if some of the lexically captured variables are tied to something else than the current buffer contents. 

I don't understand. Lexical variables are not tied to the current state, they _are_ a saved state of the past. Anyway it seems you are talking about my use case where these function's vars want to be able to reference a predictable state of the capf, recorded when candidates were first gotten. Some of that state might have been obtained from the buffer, but it's then independent from what may have changed meanwhile.

> Or if there's a hash-table using an eq test among them.

Completely lost here. For what key?

> So I'd rather not split the fix into several commits. If you don't mind, I'll amend your patch before pushing. Alternatively, I can strip the attribution.

Great!

If you like the last version you can squash the three commits and amend the commit message.

> Sorry, I didn't quire catch the meaning here.

You said a second cache introduces complexity, and I was saying that that  amount of complexity is in the problem itself, and that if you don't do anything you will eventually measure the same amount of "missing" complexity as a bug.

**Comment by dgutov** [comment]
> But in this case it is still better in practice to have some kind of optional "strict" mode of operation with which you can pressure developers without punishing users.

We'll see.

> many users assume company is the only frontend, or at least their most important one

OK, fair point.

> In my opinion, design decisions where you don't repeat actions that you determined that you don't have to repeat are good in their own right, even if accidental, and regardless of speed.

I'm saying this doesn't do enough anyway. Not in the sense that Stefan expressed as ":exit-function goes together with the completion-table" (same can be said for other properties).

> I don't understand. Lexical variables are not tied to the current state, they are a saved state of the past.

That's what I meant. Sorry if I didn't express that better. So I'm saying that stuff works if the saved state references the buffer contents, or something else that doesn't change between repeated CAPF hooks invocations, as long as the user hasn't typed anything new (or deleted something). But that's not all a backend writer might try to do.

> Completely lost here. For what key?

Completion strings. Here's the example I wanted to give:

```lisp
;; -*- lexical-binding: t -*-

(defun capf-test ()
  (let* ((completions (list "aaaaaa" "aaabbb" "aaaccc"))
         (annotations (mapcar (lambda (c) (cons c (concat " " c "z"))) completions))
         (sym (company-grab-symbol))
         (ann-fun (lambda (c) (cdr (assq c annotations)))))
    (if sym
        (list (- (point) (length sym)) (point) completions :annotation-function ann-fun))))

(add-hook 'completion-at-point-functions 'capf-test nil t)
```

Point being that using `assq` above should work and show annotations with the new code but not with the old code (where only `assoc` would work). Problem is, the original strings don't survive the `completion-all-completions` call, and the elements of that list, having had some text properties added, are not `eq` to the original strings. So this example doesn't work.

But you can probably make up an example where `annotation-function` closes over something else, like a temporary file name, or a network socket, or something else that's not referencing the current buffer contents. The fact that it's normally not very practical to do something like that has allowed the previous code to live unchanged for years.

**Comment by dgutov** [comment]
> If you like the last version you can squash the three commits and amend the commit message.

Done. In the end I've decided against doing it all in one commit. Please take a look, comments welcome.

**Comment by joaotavora** [comment]
> . Please take a look, comments welcome.

Looks great, but for a nit in the commit message.

"exit-function is a closure that references
some data dependent on the current buffer contents"

This phrase is attributed to me, but I would have said "previous buffer contents". 

Tho it's generally understood that commiters can and will edit messages, and it's not something I haven't done myself, and it's nothing super important.

**Comment by dgutov** [comment]
Yes, I've had to rewrite it so that the description is closer to "why" instead of "what". And that a year or two later I can still understand it.

"Previous" could work, but I think "current" works better, because a closure references the current buffer state when created. And also, in the scenario we're describing, a closure referencing the previous state wouldn't have been a problem. The fact that it's created anew after the completion is inserted and references the current buffer state, was the problem instead. Right?

Speaking of subsequent commits, if you have any opinion on (not) saving completion data for all buffers, indefinitely, I'm all ears.

**Comment by dgutov** [comment]
> do you have any plans for bumping the version so GNU ELPA

Sure, since you've already tried the new code, so at least it has no obvious breakages.

I'd prefer to solve that small remaining question first, though.

> If you must optimize for space

Not sure if I must, but hanging on to a reference to a piece of long-outdated data is a textbook memory leak, and that makes me uneasy.

> only save ... not potentially the completions themselves 

That would take care of the "ugly broken" case where the CAPF function is written the wrong way to being with. Which is nice, but maybe not enough. And also, this way we wouldn't be able to use the saved value in `sorted` (that one needs the completion table).

> Can buffer switches occur mid-completion that lead to mid-completion-completions?

They can (just switch to a different window, and then back). Otherwise I'd leave the variable global, and that would be the end of it. We could, of course, prohibit that from now on, but it's handy sometimes.

> If they can, you can still think of a complex scheme of detecting the interruption and restoring later.

Not sure what you mean. There basically can be as many completions going on at once as there are buffers.

> a global weak hash-tables indexed by something that you know represents a fully completed completion attempt

That is interesting, but requires framework support. For now, I'm gonna push something similar.

**Comment by joaotavora** [comment]
> That is interesting, but requires framework support. For now, I'm gonna push something similar.

Even without "framework support" you could do it by:

* setting a buffer-local `company-capf--save-current-data-key` in `candidates`
* resetting it to `nil` in `post-completion`
* using the `company-capf--save-current-key` to retrieve cached data from a new global weak hash table `company-capf--saved-data`.

This would (arguably) be simpler and more readable that constantly adding and removing things from a hook (I've used this kind of solutions in the past, but that's not really what hooks are for IMO).

**Comment by dgutov** [comment]
> resetting it to nil in post-completion

`post-completion` isn't always called. E.g. when completion is aborted, it's not.

**Comment by joaotavora** [comment]
> post-completion isn't always called. E.g. when completion is aborted, it's not.

OK, then.  Then minimal framework support for a `really-really-post-completion-even-when-aborted` maybe? :-)

**Comment by dgutov** [comment]
If you like. :) I just opted for a lightly tidier solution.

These two hooks are a bit odd, and many of their uses are likely to look like this (`company-begin-backend` is one example).

**Comment by joaotavora** [comment]
> I just opted for a slightly tidier solution.

Indeed, duh me: if you're going to use the hook, just clear the info there, no need to clear the key to any hash table.

> Which is fairly cryptic.

It's not terrible, but I would say, perhaps even more cryptically:

"CAPF :exit-function can safely reference the lexical context recorded at the moment when
the completion table is returned."

because it's up to the CAPF to decide whether that context is recorded from the buffer's contents, something in the file-system, degree of relative humidity or the position of moon.

In eglot's case the exit function needs to know the bounds of the thing-at-point when completion was invoked so it can completely undo the completion and expand a snippet in its place instead. But it could have been some information from the server, also. Perhaps you can add this small example to the NEWS, or just don't worry about it.

**Comment by monnier** [comment]
FWIW, I don't see why this deserves a NEWS entry.  It's just a bug fix.

> "CAPF :exit-function can safely reference the lexical context recorded
> at the moment when the completion table is returned."

As mentioned in some earlier message, "lexical context" is only
tangentially related: the fact that the two :exit-functions returned in
your case (the one before the completion and the one after) happen to be
closures in your particular circumstance is just happenstance.  It could
just as well be that both calls return an :exit-function that's
a plain global symbol, just not the same one.


        Stefan

**Comment by dgutov** [comment]
> FWIW, I don't see why this deserves a NEWS entry. It's just a bug fix.

It's the only change in the upcoming release. It could say "Bugfixes" (bugfix?), but I think it might be helpful for CAPF authors to know starting with which release of Company they can write exit-function's a certain way.

> It could just as well be that both calls return an :exit-function that's a plain global symbol, just not the same one.

That's a fair point, though I'm not sure how to translate it into a new entry. "Better CAPF compatibility"?

**Comment by monnier** [comment]
Maybe "Fix CAPF backend to use the right :exit-function"?

**Comment by dgutov** [comment]
Works for me, thanks!

**Comment by joaotavora** [comment]
>  happen to be closures in your particular circumstance is just happenstance. It could just as well be that both calls return an :exit-function that's a plain global symbol, just not the same one. 

Of course, but doesn't change the fact that they _can now_  be closures that reference a stable lexical context, which is something particularly useful here that you _couldn't_ do earlier.

**Comment by dgutov** [comment]
I'm happy to add a second sentence to that entry, but "lexical context" seems to be a somewhat limiting implementation detail. How about: 

> It can now safely reference (use?) data recorded at the moment when the completion table is returned

?

**Comment by monnier** [comment]
> It can now safely reference (use?) data recorded at the moment when the
> completion table is returned

Fine by me

**Comment by joaotavora** [comment]
Horrible, just horrible! Nah kidding, it was fine before, really

**Comment by dgutov** [comment]
:man_shrugging: I rewrote it again, see the attached commit.

The problem with the previous ones, I think, is that the exit-function was kind-of able to "use the data recorded when ..." already, provided it somehow managed to access it, or "safely reference the lexical context recorded when ..." if it managed to reference it at all. Just semantics, of course.
